### ISSUE TITLE
Free form images

### ISSUE THREAD

**Issue by erikmartino** [GoogleContainerTools/jib#1893]
- *Jib version:* 1.4.0
- *Build tool:* gradle
- *OS:* Mac OSX

**Description of the issue**:

jib doesn't work well for projects that are not runnable java. For example a plugin to a server.

**Expected behavior**:

A get out of the way mode, I will control the content and entrypoint

**Additional Information**: 

For example: `containerizingMode = 'none'`

**Comment by loosebazooka** [comment]
Can you explain what you want to do? You can use jib-core directory if you want to do something very free form?

**Comment by erikmartino** [comment]
My main problem was that it puts the jar file in a sub directory named classpath when using the `containerizingMode = 'packaged'` and not the app root. 

I didn't know about jib-core, I will take a look.

**Comment by erikmartino** [comment]
@loosebazooka I looked at `jib-core`, It can certainly do what I want, however I prefer to use the jib gradle tasks. I ended up including the jar file twice. One, out of harms way, in the <app-root>/classpath/xxx.jar and one where I really want it to be using the extraDirectories feature.

I believe gradle jib could benefit from a way to disable the automatic inclusion of the jar file and dependencies.

**Comment by cricket007** [comment]
> however I prefer to use the jib gradle tasks

Since Jib Core is a Java library, it can be included inside of the Grade buildscript classpath and used to serve your needs.

**Comment by polarfish** [comment]
I have a similar problem. I'm building an extension (a jar file) to a dockerized (exploded) war.
The original image is built with Jib and have the next structure:
```
app
└─ WEB-INF
    └─ lib
    └─ classes
```
The entry point is: 
```
"Entrypoint": [
    "java",
     "-cp",
     "/app/extras/*:/app/WEB-INF/classes:/app/WEB-INF/lib/*",
     "com.myapp.Application"
 ]
```
In the extension project I take the original image as `jib.from.image` preserving (inheriting) the entry point and want to put my jar to the class path, ideally to the `/app/extras/` but I cannot do that.  
- If I set `<appRoot>/app/extras/</appRoot>` and choose `<containerizingMode>packaged</containerizingMode>`, my jar goes to `/app/extras/classpath/myextension-1.0.0.jar`.  
- If I set `<appRoot>/app/extras/</appRoot>` and choose `<containerizingMode>exploded</containerizingMode>`, my classes go to `/app/extras/classes/com/myapp/...`.  
  
I ended up setting `<appRoot>/app/WEB-INF/</appRoot>` and using `<containerizingMode>exploded</containerizingMode>` to have the extensions classes merged with the original classes from the war but it's clearly a hack.  
  
It would be so nice if we had flexibility in setting the default target folder to copy to (to be able to remove these `/classpath` and `/classes` sub-paths and send the files to the place controlled with `appRoot` setting).

 PS I use **jib-maven-plugin 2.3.0**

**Comment by chanseokoh** [comment]
Hi @polarfish,

I will list a few alternatives (although I'm not sure if you'll like them).
- [Jib CLI](https://github.com/GoogleContainerTools/jib/tree/master/jib-cli#examples): a general-purpose standalone CLI too for building images. You can easily put add files.
- [Jib Core](https://github.com/GoogleContainerTools/jib/tree/master/jib-core): Java library for building images. Our Maven/Gradle plugins depend on this.
- [Jib Maven/Gradle Plugin Extensions](#2401): we are getting closer to release the plugin extension system where anyone can easily tweak and modify the way the Jib plugins assembles an image. You'll have to wait a bit more for the next release though. Initially, we'll also release several first-party extensions to address some tricky and unconventional use-cases that we do not want to support in the man Jib plugins.

**Comment by polarfish** [comment]
Hi @chanseokoh 

Thank you for the quick response. I'd wait for the plugin extension system to be ready.

**Comment by chanseokoh** [comment]
I forgot to update here that the [Jib extension framework](https://github.com/GoogleContainerTools/jib-extensions) is live.

And we have also released [Jib CLI preview](https://github.com/GoogleContainerTools/jib/tree/master/jib-cli#examples) and actively working on to make it stable.

With this set of tools, I think the original issue can be resolved in different ways. Closing the issue, but feel free to give us feedback if it's still an issue.
